### PR TITLE
Allow mappers to save subclass entities.

### PR DESCRIPTION
**Pull request by tabennett**
The STI implementation that we worked on a couple of months ago is still unusable for me without this PR; I have it implemented on the 5.1.* branch of my fork (I use it for an app I've been building at work and this solution works nicely), but we're upgrading the app to Laravel 5.3 and I need to  use the 5.3-dev branch branch and this never got merged into the original PR. We talked about this in the conversation for the previous PR: https://github.com/analogueorm/analogue/pull/131#issuecomment-245150719. Please consider merging this in, for the reason discussed in that comment thread. Also, if you want help writing tests for this feature (or anything else that I can help with) just let me know. I'm in Laravel slack as *tabennett*. Thank you for your time and work you put into this project.

### CODE REVIEW

**Comment by RemiCollin** [comment]
Sorry for late response. Do you know what cause the test to fail on this. ?

**Comment by tabennett** [comment]
Yeah, the error in the 5.6 and hhvm tests is because I added a return type hint to the `ResultBuilder::builderForResult()` method. This is my fault entirely, I'm just so used to using PHP7 now that I added it without thinking. I'm going to push out a fix that removes those. 

The php7 CI tests are failing because one of the assertions in the single table inheritance tests you wrote is failing (I'll look at those too and see if I can get them back to green).

**Comment by RemiCollin** [comment]
Merging it and having a look at those tests ;)

**Comment by tabennett** [comment]
Remi we still need to remove that return type hint. I was going to push it up to this branch this morning, but if it's easier for you to just remove it while you're in there working on that test.  It's right here:
https://github.com/analogueorm/analogue/blob/5.3-dev/src/System/ResultBuilder.php#L119

**Comment by RemiCollin** [comment]
Ok, cheers !

**Comment by RemiCollin** [comment]
I made some progress, but I face some problems with our implementation. How do you currently deal with scopes in your project ?

I want to have analogue automatically apply scope based on the Inheritance type, which was why my previous test was failing. 

Let's say we have the following structure :

Vehicle --> Car 
              --> Bicycle

We want our `VehicleMapper` to return any Car or Bicycle entities, and our CarMapper & BicycleMapper to return Car and Bicycle entities only.

I implemented a global scope based on the discriminatorMap and on the discriminatorColumn. It works on Car & Bicycle but not on the VehicleMapper (which is scoped in that case to the VehicleClass). 

And things can get even more complex if we have a deeper structure, let'say :

Vehicle --> Car --> ElectricCar
                           --> DieselCar

etc.. 

In short, we need an easy way to have analogue know which types to scope depending on the context. 

Any idea ?

**Comment by tabennett** [comment]
Right now, I'm building a pretty complex production level app using Analogue, so I"m going to use some of the examples and code I have for it.

I have a *vendor_invoices* table that holds my invoices for various trash pickup services here in the U.S. Right now, we have the following invoice types (using STI):

*VendorInvoice* - This is the base invoice class.
- *WasteManagementInvoice* - For WM invoices.
- *RepublicServicesInvoice* - For Republic Services invoices.


Here is my global scope that gets applied:
```php
class VendorInvoiceScope implements ScopeInterface
{
    /**
     * @param  Query  $query
     * @return void
     */
    public function apply(Query $query)
    {
        $entityMap = $query->getMapper()->getEntityMap();
        $type = $entityMap->type;

        if ($type !== 'Vendor Invoice') {
            $query->whereType($type);
        }
    }

    /**
     * @param  Query  $query
     * @return void
     */
    public function remove(Query $query)
    {
        // Parse query and remove type filter
    }
}
```

// Here's how I register it:
```php
$analogue->register('App\Entities\VendorInvoice\VendorInvoice', 'App\EntityMaps\VendorInvoice\VendorInvoiceMap');
$analogue->mapper('App\Entities\VendorInvoice\VendorInvoice')->addGlobalScope($vendorInvoiceScope);

$analogue->register('App\Entities\VendorInvoice\RepublicServicesInvoice', 'App\EntityMaps\VendorInvoice\RepublicServicesInvoiceMap');
$analogue->mapper('App\Entities\VendorInvoice\RepublicServicesInvoice')->addGlobalScope($vendorInvoiceScope);

$analogue->register('App\Entities\VendorInvoice\WasteManagementInvoice', 'App\EntityMaps\VendorInvoice\WasteManagementInvoiceMap');
$analogue->mapper('App\Entities\VendorInvoice\WasteManagementInvoice')->addGlobalScope($vendorInvoiceScope);
    }
```

This works very well, however I'm not doing 3 levels of nesting. If I were, I would probably have to change my scopes to be more complex. For example, if I added in *Vogel Disposal Services* I would have the following invoice types:

*VendorInvoice* - This is the base invoice class.
- *WasteManagementInvoice* - For WM invoices.
- *RepublicServicesInvoice* - For Republic Services invoices.
--*VogelDisposalServicesInvoice* - For Vogel Disposal Services invoices.

So in my global scope I'd probably have to do something like this:

```php
class VendorInvoiceScope implements ScopeInterface
{
    /**
     * @param  Query  $query
     * @return void
     */
    public function apply(Query $query)
    {
        $entityMap = $query->getMapper()->getEntityMap();
        $type = $entityMap->type;

        if ($type !== 'Vendor Invoice') {
            if ($type == 'Republic Services') {
                $query->whereType($type)
                           ->orWhereType('Vogel Disposal Services');
            } else {
                 $query->whereType($type);
            }
        }
    }

    /**
     * @param  Query  $query
     * @return void
     */
    public function remove(Query $query)
    {
        // Parse query and remove type filter
    }
}
```

That's a very hard-coded solution though and I don't know how scalable it would be in the long when (when we start adding more invoice types), but for right now it's better than nothing. 

One solution for this might be to have it also register that type filter for the entity map's parent too. For example, here's a simple approach/implementation that doesn't even need reflection (not tested though):

```php
class VendorInvoiceScope implements ScopeInterface
{
    /**
     * @param  Query  $query
     * @return void
     */
    public function apply(Query $query)
    {
        $entityMap = $query->getMapper()->getEntityMap();
        $type = $entityMap->type;

        // The Vendor Invoice Mapper should return ALL vendor invoice types, so no filter is added if the entity map is VendorInvoiceMap
        if ($type !== 'Vendor Invoice') {
            $query->whereType($type);
            $parentClass = get_parent_class($entityMap);

            if ($parentClass !== 'App\EntityMaps\VendorInvoiceMap') {
                // This entity map extends from another besides the base VendorInvoiceMap, so we'll add a filter for it too.
                $properties = get_class_vars($parentClass);
                $discriminatorColumn = $properties['discriminatorColumn'];
                $query->orWhereType($properties[$discriminatorColumn]);
            }
        }
    }

    /**
     * @param  Query  $query
     * @return void
     */
    public function remove(Query $query)
    {
        // Parse query and remove type filter
    }
}
```

If that works, it would be really useful to abstract it away into a simpler function that can just be called on an entity map or something and it will automatically take care of the STI scoping for it.

Aside from this, it's also pretty cumbersome to register entity events too. Here's how I register my model events right now:

```php
$setPriority = function (VendorInvoice $invoice) {
    if (!$invoice->canBeAllocated()) {
        $invoice->priority = VendorInvoice::PRIORITY_UNKNOWN;
    } else {
        if ($invoice->extraneous_charges->count() > 0
            || $invoice->rolloff_service_charges->count() > 0
            || $invoice->pass_through_service_charges->count() > 0) {
            $invoice->priority = VendorInvoice::PRIORITY_HIGH;
        } else {
            $invoice->priority = VendorInvoice::PRIORITY_LOW;
        }
    }

    return true;
};

$toggleDeployable = function(VendorInvoice $invoice) {
    if ($invoice->isDirty('marked_as_deployable_at') && $invoice->priority == 2) {
        if ($invoice->marked_as_deployable_at !== null) {
            dispatch(new SendInvoiceToFeeScheduler($invoice));
        } else {
            dispatch(new RemoveInvoiceFromFeeScheduler($invoice));
        }
    }
};

$analogue->mapper(WasteManagementInvoice::class)->registerEvent('storing', $setPriority);
$analogue->mapper(RepublicServicesInvoice::class)->registerEvent('storing', $setPriority);

$analogue->mapper(WasteManagementInvoice::class)->registerEvent('stored', $toggleDeployable);
$analogue->mapper(RepublicServicesInvoice::class)->registerEvent('stored', $toggleDeployable);
```

Ideally I'd like to be able to just say: `$analogue->mapper(VendorInvoice::class)->registerEvent('storing', $setPriority);` and have it work (without having to register the listener on each subclass type). But this isn't currently that big of an issues for me (yet).

**Comment by tabennett** [comment]
Wait, that query scope is backwards; When we have a child class, we don't want to fetch parent type classes with it. We would want to fetch all child class records when we fetch a parent class. Let me think about this for a bit.

**Comment by RemiCollin** [comment]
Thanks for your detailed answer ! I think that a possible solution could be to have the scope parse all the registered entity classes in analogue, and add any class that is a subclass of the scoped entity to the query constraint. The nice thing with approach is that the parent doesn't need to know about its subclass.

Of course, that's assuming that all Entity class are explicitely registered and that all registration has happened before any mapper is instantiated. I'm pretty sure that's not a big deal though, as I added a `strict mode`some times ago and by turning it to true we can prevent this behavior by throwing an exception when instantiating any non-registered Mapper.

I'll implement this and see how it goes.

And yes the events will need to react to this new feature in the same fashion, thanks for pointing this out!

**Comment by RemiCollin** [comment]
Well, seems to work quite good. I pushed the implementation on 5.3-dev

**Comment by tabennett** [comment]
Man this is really slick! This is pretty much exactly what I had in mind (except I had it backwards), but I figured we'd need a way to parse through the registered mappers and apply the scopes, etc. I just pulled down the latest release 5.3.0 into my app and it's working like beautifully! Thanks again for helping me get this in here man and thank for Analogue!

**Comment by RemiCollin** [comment]
Well, i'm happy to have this finally working, as it's been a feature I wanted for quite a long time. Still have to address this event stuff as well, probably with the same approach
